libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt

```diff
@@ -0,0 +1,237 @@
+    var onCvvInfoClicked: (() -> Unit)? = null
+    var onCardNumberComplete: ((Boolean) -> Unit)? = null
+    var onCvvComplete: ((Boolean) -> Unit)? = null
+    var openMonthSelectionListener: (() -> Unit)? = null
+    var openYearSelectionListener: (() -> Unit)? = null
+
+    private val binding: ViewCardInputBinding = inflate(R.layout.view_card_input)
+
+    private val validator by lazy { CreditCardValidator() }
+
+    constructor(context: Context) : super(context)
+
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
+        context,
+        attrs,
+        defStyleAttr
+    )
+
+    init {
+        if (isInEditMode) View.inflate(context, R.layout.view_card_input, this)
+
+        setUpView()
+    }
+
+    /**
+     *
+     * Sets the view fields with given [CardInputViewState].
+     *
+     * @param viewState is [CardInputViewState].
+     */
+    fun setViewState(viewState: CardInputViewState?) {
+        viewState?.run {
+            binding.viewState = this
+            binding.executePendingBindings()
+        }
+    }
+
+    /**
+     *
+     * Validates all card fields and if fields are not valid,
+     * it will immediately sets invalid field backgrounds with [CardInputViewState.errorBackgroundColor].
+     *
+     * @return [Boolean.true] if all fields are valid, [Boolean.false] if one or more field is invalid.
+     */
+    fun validate(): Boolean {
+        val cardInformation = binding.viewState?.cardInformation ?: CardInformation()
+
+        val isCardNumberValid = validator.isCardNumberValid(cardInformation.cardNumber)
+        val isExpiryMonthValid = validator.isExpiryMonthValid(cardInformation.expiryMonth)
+        val isExpiryYearValid = validator.isExpiryYearValid(cardInformation.expiryYear)
+        val isCvvValid = validator.isCvvValid(cardInformation.cvv)
+
+        binding.viewState = binding.viewState?.copy(
+            cardNumberValid = isCardNumberValid,
+            expiryMonthValid = isExpiryMonthValid,
+            expiryYearValid = isExpiryYearValid,
+            cvvValid = isCvvValid
+        )
+        binding.executePendingBindings()
+
+        return isCardNumberValid && isExpiryMonthValid && isExpiryYearValid && isCvvValid
+    }
+
+    /**
+     *
+     * Validates all card fields and if fields are not valid,
+     * it will immediately sets invalid field backgrounds with [CardInputViewState.errorBackgroundColor].
+     *
+     * @see validate
+     *
+     * @return [CardInformation] if all fields are valid, if not returns null.
+     */
+    fun validateEndGet(): CardInformation? =
+        if (validate()) binding.viewState?.cardInformation else null
+
+    /**
+     *
+     * Resets [CardInputViewState] data including [CardInputViewState.cardInformation] and
+     * validity data.
+     */
+    fun reset() {
+        binding.viewState = binding.viewState?.reset()
```

Comment:
Generating a new ViewState without any copying process is not the responsibility of view state object.
I mean we need a secondary constructor to generate an empty view state.

Reply:
In `reset()`, my goal is keep the initial configurations for `CardInputView`; textColor, backgrounds etc. I tried to achive configuration save without storing it in another state holder or in builder.